src/metaxy/metadata_store/vortex.py

```diff
@@ -0,0 +1,315 @@
+            )
+
+        # Local path (including file:// and local:// URLs)
+        if root_str.startswith("file://"):
+            root_str = root_str[7:]
+        elif root_str.startswith("local://"):
+            root_str = root_str[8:]
+        local_path = Path(root_str).expanduser().resolve()
+        self._root_uri = str(local_path)
+
+        super().__init__(
+            fallback_stores=fallback_stores,
+            versioning_engine_cls=PolarsVersioningEngine,
+            versioning_engine="polars",
+            **kwargs,
+        )
+
+    # ===== MetadataStore abstract methods =====
+
+    def _has_feature_impl(self, feature: CoercibleToFeatureKey) -> bool:
+        """Check if feature exists in Vortex store."""
+        feature_key = self._resolve_feature_key(feature)
+        return self._table_exists(self._feature_uri(feature_key))
+
+    def _get_default_hash_algorithm(self) -> HashAlgorithm:
+        """Use XXHASH64 by default to match other non-SQL stores."""
+        return HashAlgorithm.XXHASH64
+
+    @contextmanager
+    def _create_versioning_engine(
+        self, plan: FeaturePlan
+    ) -> Iterator[PolarsVersioningEngine]:
+        """Create Polars versioning engine for Vortex store."""
+        with self._create_polars_versioning_engine(plan) as engine:
+            yield engine
+
+    @contextmanager
+    def open(self, mode: AccessMode = "read") -> Iterator[Self]:  # noqa: ARG002
+        """Open the Vortex store.
+
+        Creates the root directory if it doesn't exist.
+        Vortex files are opened on-demand per operation.
+
+        Args:
+            mode: Access mode (accepted for API consistency).
+
+        Yields:
+            Self: The store instance with connection open
+        """
+        self._context_depth += 1
+
+        try:
+            if self._context_depth == 1:
+                Path(self._root_uri).mkdir(parents=True, exist_ok=True)
+                self._is_open = True
+                self._validate_after_open()
+
+            yield self
+        finally:
+            self._context_depth -= 1
+            if self._context_depth == 0:
+                self._is_open = False
+
+    # ===== Internal helpers =====
+
+    def _feature_uri(self, feature_key: FeatureKey) -> str:
+        """Return the path used by Vortex for this feature."""
+        if self.layout == "nested":
+            table_path = "/".join(part for part in feature_key.parts if part)
+        else:
+            table_path = feature_key.table_name
+        return f"{self._root_uri}/{table_path}.vortex"
+
+    def _table_exists(self, table_uri: str) -> bool:
+        """Check whether the provided path contains a Vortex file."""
+        try:
+            import vortex
+
+            _ = vortex.open(table_uri)
+            return True
+        except (FileNotFoundError, OSError, Exception):
+            return False
+
+    # ===== Storage operations =====
+
+    def write_metadata_to_store(
```

Comment:
just a file format - not a table format like ducklake, iceberg, delta. This means some extra complexity

Reply:
https://spiceai.org/blog/releases/v1.9.0 Cayenne Data Accelerator  could be interesting here

Reply:
and https://github.com/spiraldb/iceberg something in JVM land